core/store/src/trie/trie_storage.rs

```diff
@@ -263,22 +263,23 @@ pub trait TrieStorage {
 pub struct TrieRecordingStorage {
     pub(crate) store: Store,
     pub(crate) shard_uid: ShardUId,
-    pub(crate) recorded: RefCell<HashMap<CryptoHash, Vec<u8>>>,
+    pub(crate) recorded: RefCell<HashMap<CryptoHash, Arc<[u8]>>>,
 }
 
 impl TrieStorage for TrieRecordingStorage {
     fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Arc<[u8]>, StorageError> {
         if let Some(val) = self.recorded.borrow().get(hash) {
-            return Ok(val.as_slice().into());
+            return Ok(val.clone());
```

Comment:
I am very slightly in favor of spelling this as `Arc::clone(val)`, to make it clear there's no memcpy here. OTOH, I'd probably even write this as `self.recorded.borrow().get(hash).cloned()` anyway :)